usr/initiator.c

```diff
@@ -2050,12 +2051,19 @@ iscsi_sync_session(node_rec_t *rec, queue_task_t *qtask, uint32_t sid)
 		goto destroy_session;
 	}
 
-	session->r_stage = R_STAGE_SESSION_REOPEN;
-
 	err = sync_conn(session, 0);
 	if (err)
 		goto destroy_session;
 
+	iscsi_sysfs_get_session_target_unbound(&target_unbound, sid);
+
+	if (target_unbound) {
+		log_error("Shutdown unbound session%d", sid);
+		session_conn_shutdown(&session->conn[0], qtask, 0);
```

Comment:
The conn->state will be ISCSI_CONN_STATE_CLEANUP_WAIT from sync_conn() right?

If so, the conn could have been in the logged in state when the unbind was running and iscsid stopped. So we still need to do a stop_conn call so the kernel can run libiscsi.c:iscsi_conn_stop.

I think we can check the conn kernel state to detect if it's in the DOWN state already (scsi_transport_iscsi.c:iscsi_stop_conn sets that state when userspace has called into the kernel to stop a conn for removal).

Reply:
> The conn->state will be ISCSI_CONN_STATE_CLEANUP_WAIT from sync_conn() right?
> 

Yes, it's now ISCSI_CONN_STATE_CLEANUP_WAIT state.

> If so, the conn could have been in the logged in state when the unbind was running and iscsid stopped. So we still need to do a stop_conn call so the kernel can run libiscsi.c:iscsi_conn_stop.
> 

Yes, we need to check it and judge if stop_conn is necessary now.

> I think we can check the conn kernel state to detect if it's in the DOWN state already (scsi_transport_iscsi.c:iscsi_stop_conn sets that state when userspace has called into the kernel to stop a conn for removal).